src/action/network.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2020 Google LLC
+//
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+//! A handler and associated types for the network connection action.
+//!
+//! The network connection action lists all the opened connections on the
+//! client and the information about them (e.g. type of the connection, source
+//! and target IP addresses, the process that owns the connection)
+
+use std::fmt::{Display, Formatter};
+use std::net::IpAddr;
+
+use log::error;
+use sysinfo::{System, SystemExt, Process, ProcessExt};
+use netstat2::{
+    self,
+    ProtocolSocketInfo::{self, Tcp, Udp},
+    TcpState
+};
+use rrg_proto::{
+    ListNetworkConnectionsArgs,
+    NetworkConnection,
+    NetworkEndpoint,
+    network_connection::{Family, Type, State}
+};
+
+use crate::session::{self, Session};
+
+#[derive(Debug)]
+struct Error {
+    connection_info_error: netstat2::error::Error,
+}
+
+impl std::error::Error for Error {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.connection_info_error)
+    }
+}
+
+impl Display for Error {
+
+    fn fmt(&self, fmt: &mut Formatter) -> std::fmt::Result {
+        write!(
+            fmt,
+            "failed to list network connections: {}",
+            self.connection_info_error
+        )
+    }
+}
+
+impl From<netstat2::error::Error> for Error {
+
+    fn from(error: netstat2::error::Error) -> Error {
+        Error { connection_info_error: error }
+    }
+}
+
+/// A request type for the network connections action.
+pub struct Request {
+    listening_only: bool
+}
+
+/// A type that holds brief information about the process.
+struct ProcessInfo {
+    /// Process ID.
+    pid: u32,
+    /// Process name.
+    name: Option<String>
+}
+
+/// A response type for the network connections action.
+pub struct Response<'a> {
+    /// Information about the socket which is used by connection.
+    socket_info: &'a ProtocolSocketInfo,
+    /// Information about the process that owns the connection.
+    process_info: Option<ProcessInfo>
+}
+
+/// Gets the IP address type from a given address.
+fn make_family(addr: &IpAddr) -> Family {
+    match addr {
+        IpAddr::V4(_) => Family::Inet,
+        IpAddr::V6(_) => Family::Inet6
+    }
+}
+
+/// Contructs NetworkEndpoint from the specified address and port.
+fn make_network_endpoint(addr: &IpAddr, port: u16) -> NetworkEndpoint {
+    NetworkEndpoint {
+        ip: Some(addr.to_string()),
+        port: Some(port as i32)
+    }
+}
+
+/// Converts [netstat2::TcpState][tcp_state] to `State`
+/// enum used in the protobuf
+///
+/// [tcp_state]: ../../../netstat2/enum.TcpState.html
+fn make_state(state: &TcpState) -> State {
+    match state {
+        TcpState::Unknown     => State::Unknown,
```

Comment:
These aligned arrows look nice, but they are bad for diffs—if a new enum value is introduced with a longer name, you have to re-align everything, which is not great. For this reason I prefer not to do them.